### PR TITLE
Fixed changing fingerprints even though training data did not change

### PR DESCRIPTION
**Pull request by tmbo**
**Issue**:
- we relied on hashes in some places to fingerprint parts of the configuration / training data. hashes of strings are not stable across python (>3.3) runs, which means that the fingerprint changes even though the configuration / data is still the same

**Proposed changes**:
- implement proper fingerprinting using md5 for strings instead of python native hashes

**Status (please check what you already did)**:
- [x] added some tests for the functionality
- [x] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/master/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

File: tests/utils/test_io.py [hunk 1]
```diff
@@ -91,3 +93,34 @@ def test_write_utf_8_yaml_file(tmp_path: Path):
...
+        [],
+        ["a"],
+        [{}],
+        [None],
+    ],
+)
+def test_fingerprint_dicionaries(dictionary):
+    assert rasa.shared.utils.io.deep_container_fingerprint(
+        dictionary
+    ) == rasa.shared.utils.io.deep_container_fingerprint(copy.deepcopy(dictionary))
```
Comment (tmbo) [inline tests/utils/test_io.py:115]:
```suggestion
@pytest.mark.parametrize(
    "container",
    [
        {},
        {"hello": "world"},
        {1: 2},
        {"foo": ["bar"]},
        {"a": []},
        [],
        ["a"],
        [{}],
        [None],
    ],
)
def test_fingerprint_containers(container):
    assert rasa.shared.utils.io.deep_container_fingerprint(
        container
    ) == rasa.shared.utils.io.deep_container_fingerprint(copy.deepcopy(container))
```

File: rasa/shared/utils/io.py [hunk 2]
```diff
@@ -194,6 +194,75 @@ def list_subdirectories(path: Text) -> List[Text]:
...
+
+    The dictionary can contain any keys and values which are either a dict,
+    a list or a elements which can be dumped as a string.
+
+    Args:
+        dictionary: dictionary to be hashed
+        encoding: encoding used for dumping objects as strings
+
+    Returns:
+        the hash of the dictionary
```
Comment (alwx) [inline rasa/shared/utils/io.py:234]: We usually start from a capital letter here.

Comment (alwx) [reply rasa/shared/utils/io.py:234]: But it's probably tested automatically anyway :)

**Review by alwx — APPROVED** [review]
Looks great! 🚀

**Pull request merged**